src/runtime-rs/crates/runtimes/src/tracer.rs

```diff
@@ -0,0 +1,164 @@
+pub fn trace_setup(
+    sid: &str,
+    jaeger_endpoint: &str,
+    jaeger_username: &str,
+    jaeger_password: &str,
+) -> Result<()> {
+    let mut kt = KATATRACER.lock().unwrap();
+
+    // If varify jaeger config returns an error, it means that the tracing should not be enabled
+    let endpoint = varify_jaeger_config(jaeger_endpoint, jaeger_username, jaeger_password)?;
+
+    // enable tracing
+    kt.enable();
+
+    // derive a subscriber to collect span info
+    let tracer = opentelemetry_jaeger::new_collector_pipeline()
+        .with_service_name(format!("kata-sb-{}", &sid[0..min(8, sid.len())]))
+        .with_endpoint(endpoint)
+        .with_username(jaeger_username)
+        .with_password(jaeger_password)
+        .with_hyper()
+        .install_batch(Tokio)?;
+
+    let layer = tracing_opentelemetry::layer().with_tracer(tracer);
+
+    let sub = Registry::default().with(layer);
+
+    // we use Arc to let global subscriber and katatracer to SHARE the SAME subscriber
+    // this is for record the global subscriber into a global variable KATATRACER for more usages
+    let subscriber = Arc::new(sub);
+    tracing::subscriber::set_global_default(subscriber.clone())?;
+    kt.subscriber = subscriber;
+
+    info!(sl!(), "Tracing enabled successfully");
+    Ok(())
+}
+
+/// Global function to shutdown the tracer and emit the span info to jaeger agent
+/// The tracing information is only partially update to jaeger agent before this function is called
+pub fn trace_end() {
+    if KATATRACER.lock().unwrap().enabled() {
+        global::shutdown_tracer_provider();
+    }
+}
+
+/// Enter the global ROOTSPAN, called at once if config.runtime.entracing is set
+/// This function is a hack on tracing library's guard approach, letting the span
+/// to enter without using a RAII guard to exit. This function should only be called
+/// once, and also in paired with [`trace_exit_root()`].
+pub fn trace_enter_root() {
+    enter(&ROOTSPAN);
+}
+
+/// Exit the global ROOTSPAN, called when shutdown request is sent by containerd and the
+/// shim actually does decide to shutdown the sandbox. This should be called in paired with [`trace_enter_root()`].
+pub fn trace_exit_root() {
+    exit(&ROOTSPAN);
+}
+
+/// let the subscriber enter the span, this has to be called in pair with exit(span)
+/// This function allows **cross function span** to run without span guard
+fn enter(span: &Span) {
+    let kt = KATATRACER.lock().unwrap();
+    let id: Option<span::Id> = span.into();
+    kt.subscriber.enter(&id.unwrap());
+}
+
+/// let the subscriber exit the span, this has to be called in pair to enter(span)
+fn exit(span: &Span) {
+    let kt = KATATRACER.lock().unwrap();
+    let id: Option<span::Id> = span.into();
+    kt.subscriber.exit(&id.unwrap());
+}
+
+/// Varifying the configuration of jaeger and setup the default value
+fn varify_jaeger_config(endpoint: &str, username: &str, passwd: &str) -> Result<String> {
```

Comment:
This looks like it might a typo ("verify"?):

```suggestion
/// Verifying the configuration of jaeger and setup the default value
fn verify_jaeger_config(endpoint: &str, username: &str, passwd: &str) -> Result<String> {
```

Reply:
Fixed.